primitives/src/vrf.rs

```diff
@@ -232,3 +250,170 @@ pub mod sortition {
+	let transcript = make_transcript(transcript_data(pre_hash, epoch, task_id));
+	pubkey
+		.vrf_verify(transcript, &output, &proof)
+		.map(|(inout, _proofbatchable)| inout)
+}
+
+pub fn finalize_randomness(
+	public: &PublicKey,
+	transcript: Transcript,
+	output: &VRFOutput,
+) -> Option<VRFInOut> {
+	match output.attach_input_hash(public, transcript) {
+		Ok(inout) => Some(inout),
+		Err(e) => {
+			log::warn!(
+				target = "VRF",
+				"attaching transcript failed due to {e}, the random is unusable."
+			);
+			None
+		},
+	}
+}
+
+#[runtime_interface]
+pub trait Vrf {
+	fn generate_vrf(
+		&mut self,
+		key_type_id: KeyTypeId,
+		pubkey: &Public,
+		pre_hash: H256,
+		epoch: u64,
+		task_id: H256,
+	) -> Option<(VRFOutput, VRFProof)> {
+		let keystore = &***self
+			.extension::<KeystoreExt>()
+			.expect("No `keystore` associated for the current context!");
+		let public_data = transcript_data(pre_hash, epoch, task_id);
+		match SyncCryptoStore::sr25519_vrf_sign(keystore, key_type_id, pubkey, public_data) {
+			Ok(Some(signature)) => {
+				let VRFSignature { output, proof } = signature;
+				Some((VRFOutput(output), VRFProof(proof)))
+			},
+			Ok(None) => {
+				log::warn!(target = "VRF", "missing Public {pubkey} from {key_type_id:?}");
+				None
+			},
+			Err(e) => {
+				log::error!(target = "VRF", "TODO: VRF signing failed {e}!");
+				None
+			},
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::vrf::generate_vrf;
+	use super::*;
+	use runtime_utils::ExtBuilder;
+	use sp_consensus_vrf::schnorrkel::PublicKey;
+	use sp_core::blake2_256;
+
+	struct PublicData {
+		key_type_id: KeyTypeId,
+		pre_hash: H256,
+		epoch: u64,
+		task_id: H256,
+	}
+
+	fn mocked_public_data() -> PublicData {
+		let key_type_id = KeyTypeId(*b"gots");
+		let pre_hash = blake2_256(b"predecessor hash").into();
+		let epoch = 1u64;
+		let task_id = blake2_256(b"task").into();
+		PublicData { key_type_id, pre_hash, epoch, task_id }
+	}
+
+	fn add_testing_key(key_type_id: KeyTypeId, builder: &ExtBuilder<()>) -> Public {
+		builder
+			.keystore
+			.as_ref()
+			.expect("A keystore")
+			.sr25519_generate_new(key_type_id, Some("//fixed"))
+			.unwrap()
+	}
+
+	#[test]
+	fn generate_vrf_output_should_be_different_for_different_private_data() {
+		let PublicData { key_type_id: keyring_id, pre_hash, epoch, task_id } = mocked_public_data();
+
+		let builder = ExtBuilder::<()>::default().with_keystore();
+		let pubkey_struct = add_testing_key(keyring_id, &builder);
+
+		builder.build_sans_config().execute_with(|| {
+			let (mut previous_public_seed, _proof) =
+				generate_vrf(keyring_id, &pubkey_struct, pre_hash, epoch, task_id).unwrap();
+			let pubkey = PublicKey::from_bytes(&pubkey_struct.0).unwrap();
+			let transcript = make_transcript(transcript_data(pre_hash, epoch, task_id));
+			let mut previous_seed =
+				finalize_randomness(&pubkey, transcript, &previous_public_seed).unwrap();
+
+			assert_ne!(previous_public_seed.0, previous_seed.to_output());
```

Comment:
@CAGS295 this assertion fails. Same on line 391.

I was under the impression that seeds will be different after each step. Am I missing something ?

Reply:
The public data is the same, the signer is the same, `previous_seed.to_output()` is in fact a copy of `previous_public_seed.0`.